### ISSUE TITLE
policy/v1beta1 no longer available in K8S 1.25

### ISSUE THREAD

**Issue by andysnowden** [DataDog/datadog-operator#619]
**Output of the info page (if this is a bug)**
```
{"level":"ERROR","ts":"2022-09-28T19:38:58Z","logger":"setup","msg":"Problem running manager","error":"failed to wait for datadogagent caches to sync: no matches for kind \"PodDisruptionBudget\" in version \"policy/v1beta1\""}
```

**Describe what happened:**
The operator is failing due to a deprecated API version.

**Describe what you expected:**
The operator to start

**Steps to reproduce the issue:**
Helm install operator on a 1.25.x cluster

**Additional environment details (Operating System, Cloud provider, etc):**
KOPS deployed cluster on AWS

This appears to be the only non `v1` used in the code base.  The `v1` has been available since 1.21, and that version of K8S is already marked as EOL.  It does mean limiting the backward support of the operator to some extent.

**Comment by celenechang** [comment]
Hi @andysnowden , thanks for reporting - it's on our radar and we'll share an update here when it's addressed!

**Comment by andysnowden** [comment]
Great to hear @celenechang 

Is there a rough ETA or potential workaround until then?

**Comment by Xosmond** [comment]
Looks like this blocks me from installing operator on my cluster. If anyone know a way to temporary patch or a workaround would be much appreciated. Thanks.

**Comment by diegoparrilla** [comment]
Any update on this topic?

**Comment by renchap** [comment]
This is blocking us as well, any ETA?

**Comment by levan-m** [comment]
Hello, we merged two changes #683, #688 addressing this issue, they will be included in the next v1.0.0 RC.

**Comment by dlorent** [comment]
This is blocking us as well - looking forward to the release.

**Comment by joe-carpenter** [comment]
also blocked here - any updates?

**Comment by levan-m** [comment]
The change is available in `v1.0.0-RC7` and later. We haven't backported it to `v0.8` yet, since we are prioritizing `v1.0` GA. It's in our backlog and will post an ETA when we get to it.

As a general ask - please, indicate the version you are using when reporting an issue.

**Comment by NissesSenap** [comment]
@levan-m the problem is that we can't even try v1.0.0-RC8 since there isn't any helm release for it.
If you could release v1.0.0-RC8 as a helm chart and I think the majority of us will be happy to give it a try just so we can upgrade to 1.25.
You would also get a win from this by getting more people to try your RC before 1.0.0 is out.
If you can prioritize this it would be great.

Or of course backport to an old version.

**Comment by levan-m** [comment]
There is not (and won't be) a new chart for 1.0. Existing Operator [chart](https://github.com/DataDog/helm-charts/tree/main/charts/datadog-operator) can be configured to install a specific version of Operator. It's now set to 0.8.4 [here](https://github.com/DataDog/helm-charts/blob/main/charts/datadog-operator/values.yaml#L45) and will be updated to 1.0 once it's generally available.

Please keep in mind, when installing Operator 1.0 with above chart you need to set CRD version explicitly to `v2alpha1` using `datadog-crds.migration.datadogAgents.version`.  Please, check issue [689](https://github.com/DataDog/datadog-operator/issues/689), specifically [this comment](https://github.com/DataDog/datadog-operator/issues/689#issuecomment-1402761227) about 1.0 setup.

**Comment by BrokenFlame** [comment]
@levan-m The workaround throws new errors and the operator still cannot start:

{"level":"ERROR","ts":"2023-03-19T07:35:53Z","logger":"klog","msg":"Observed a panic: \"invalid memory address or nil pointer dereference\" (runtime error: invalid memory address or nil pointer dereference)\ngoroutine 666 [running]:\nk8s.io/apimachinery/pkg/util/runtime.logPanic({0x195e040?, 0x2ce1c60})\n\t/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/runtime/runtime.go:74 +0x86\nk8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x73?})\n\t/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/runtime/runtime.go:48 +0x75\npanic({0x195e040, 0x2ce1c60})\n\t/usr/local/go/src/runtime/panic.go:838 +0x207\ngithub.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).getCredentialsV2(0xc0003acc30?, 0xc0003acb00)\n\t/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:611 +0x69\ngithub.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).setupV2(0xc000437dc0)\n\t/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:261 +0x24f\ngithub.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).connectToDatadogAPI(0xc000437dc0)\n\t/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:303 +0x58\nk8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x18, 0xc000680800})\n\t/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:220 +0x1b\nk8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x1e887b0?, 0xc000838680?}, 0xc0001d5dd0?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:233 +0x57\nk8s.io/apimachinery/pkg/util/wait.poll({0x1e887b0, 0xc000838680}, 0x10?, 0xf4a905?, 0xc0009fe570?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:580 +0x38\nk8s.io/apimachinery/pkg/util/wait.PollImmediateUntilWithContext({0x1e887b0, 0xc000838680}, 0x10?, 0xc000680800?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:545 +0x49\nk8s.io/apimachinery/pkg/util/wait.PollImmediateUntil(0xc000437dc0?, 0xc000a42480?, 0x0?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:536 +0x7c\ngithub.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).start(0xc000437dc0, 0x0?)\n\t/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:165 +0x10c\ncreated by github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*ForwardersManager).Register\n\t/workspace/pkg/controller/utils/datadog/forwarders_manager.go:69 +0x365\n"}
29
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
28
	panic: runtime error: invalid memory address or nil pointer dereference
27
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13e5029]
26
25
goroutine 666 [running]:
24
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x73?})
23
	/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/runtime/runtime.go:55 +0xd8
22
panic({0x195e040, 0x2ce1c60})
21
	/usr/local/go/src/runtime/panic.go:838 +0x207
20
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).getCredentialsV2(0xc0003acc30?, 0xc0003acb00)
19
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:611 +0x69
18
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).setupV2(0xc000437dc0)
17
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:261 +0x24f
16
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).connectToDatadogAPI(0xc000437dc0)
15
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:303 +0x58
14
k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x18, 0xc000680800})
13
	/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:220 +0x1b
12
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x1e887b0?, 0xc000838680?}, 0xc0001d5dd0?)
11
	/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:233 +0x57
10
k8s.io/apimachinery/pkg/util/wait.poll({0x1e887b0, 0xc000838680}, 0x10?, 0xf4a905?, 0xc0009fe570?)
9
	/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:580 +0x38
8
k8s.io/apimachinery/pkg/util/wait.PollImmediateUntilWithContext({0x1e887b0, 0xc000838680}, 0x10?, 0xc000680800?)
7
	/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:545 +0x49
6
k8s.io/apimachinery/pkg/util/wait.PollImmediateUntil(0xc000437dc0?, 0xc000a42480?, 0x0?)
5
	/go/pkg/mod/k8s.io/apimachinery@v0.23.5/pkg/util/wait/wait.go:536 +0x7c
4
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).start(0xc000437dc0, 0x0?)
3
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:165 +0x10c
2
created by github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*ForwardersManager).Register
1
	/workspace/pkg/controller/utils/datadog/forwarders_manager.go:69 +0x365

**Comment by mikesplain** [comment]
We're running into the above panic as well, any workaround @levan-m? Thanks for your work on this.

**Comment by mikesplain** [comment]
That's right, I was using v1alpha1 but was assuming either the webhook or the `migration.datadogAgents.version=v2alpha1` would take care of the migration for us or a better error then a panic. 

Bummer that we're kind of blocked on this but I understand your teams priority to get 1.0 out the door. Any pointers to manually migrating to v2alpha1 or WIP code on webhook/docs? Thanks

**Comment by levan-m** [comment]
I am sorry you getting blocked and for the confusing naming in our chart! 

Webhook is controlled by a different property in the Helm chart (see [here](https://github.com/DataDog/helm-charts/blob/main/charts/datadog-operator/values.yaml#L101-L117)), but actually enabling it is more than flipping that flag.

To unblock yourself you could take a look at this doc update PR https://github.com/DataDog/datadog-operator/pull/665, this may be helpful if you try to manually map `v1alpha1` manifest to `v2alpha1`.

This [DataDog/helm-chart#918](https://github.com/DataDog/helm-charts/pull/918/files) adds webhook configuration and documentation to the Chart. However it still has several issues and may not work out of the box.

**Comment by OurFriendIrony** [comment]
I'm also currently blocked by this issue.
Is this still on track for a mid-April release?

**Comment by pythian-chaitanya** [comment]
We are also blocked with this issue.
Any update on the release?

**Comment by levan-m** [comment]
We released `1.0.0` about two weeks ago and recommend migrating to it to avoid this issue. Right now we don't have plans to backport the fix in `0.x.x` versions.


The is the official migration guide
https://docs.datadoghq.com/containers/guide/datadogoperator_migration/

Please also check out update Operator and Helm Chart readme docs
https://github.com/DataDog/datadog-operator/blob/main/README.md
https://github.com/DataDog/helm-charts/blob/main/charts/datadog-operator/README.md

Given this problem has been addressed since `1.0.0-rc7` and we have a migration path to `1.0.0`, I consider the issue resolved.
